File: app/src/main/java/me/anon/lib/handler/ExceptionHandler.java
package me.anon.lib.handler;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Environment;

import java.io.File;
import java.io.FilenameFilter;

public class ExceptionHandler
{
	public static String VERSION = "";
	public static String VERSION_CODE = "";
	public static String PACKAGE_NAME = "";

	private String filesPath = "/";
	private String[] stackTraceFileList = null;

	private static ExceptionHandler instance;

	public static ExceptionHandler getInstance()
	{
		if (instance == null)
		{
			synchronized (ExceptionHandler.class)
			{
				if (instance == null)
				{
					instance = new ExceptionHandler();
				}
			}
		}

		return instance;
	}

	public String getFilesPath()
	{
		return filesPath;
	}

	/**
	 * @param context
	 */
	public void register(Context context)
	{
		PackageManager pm = context.getPackageManager();

		try
		{
			PackageInfo pi = pm.getPackageInfo(context.getPackageName(), 0);
			PACKAGE_NAME = pi.packageName;
			VERSION = String.valueOf(pi.versionName);
			VERSION_CODE = String.valueOf(pi.versionCode);
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}

		filesPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).getPath() + "/GrowTracker/crashes/";
		new File(filesPath).mkdirs();

		new Thread()
		{
			@Override public void run()
			{
				UncaughtExceptionHandler currentHandler = Thread.getDefaultUncaughtExceptionHandler();

				// don't register again if already registered
				if (!(currentHandler instanceof DefaultExceptionHandler))
				{
					Thread.setDefaultUncaughtExceptionHandler(new DefaultExceptionHandler(currentHandler));
				}
			}
		}.start();
	}

	/**
	 * Search for stack trace files.
	 *
	 * @return
	 */
	public String[] searchForStackTraces()
	{
		if (stackTraceFileList != null)
		{
			return stackTraceFileList;
		}

		File dir = new File(filesPath);

		// Try to create the files folder if it doesn't exist
		dir.mkdir();

		FilenameFilter filter = new FilenameFilter()
		{
			@Override public boolean accept(File dir, String name)
			{
				return name.endsWith(".stacktrace");
			}
		};

		return (stackTraceFileList = dir.list(filter));
	}

	/**
	 * Forces a manual exception post
	 * @param e
	 */
	public void sendException(Exception e)
	{
		DefaultExceptionHandler.sendException(e, "CAUGHT EXCEPTION");
	}
}

File: app2/src/main/java/me/anon/grow3/util/handler/ExceptionHandler.kt
package me.anon.grow3.util.handler

import android.content.Context
import java.io.File
import java.io.FilenameFilter
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ExceptionHandler @Inject constructor(context: Context)
{
	companion object
	{
		var VERSION = ""
		var VERSION_CODE = ""
		var PACKAGE_NAME = ""
	}

	private var exceptionHandler: DefaultExceptionHandler? = null
	public val filesPath: String = File(context.filesDir.absolutePath + "/crashreports/").apply {
		mkdirs()
	}.absolutePath

	init {
		val pm = context.packageManager
		try
		{
			val pi = pm.getPackageInfo(context.packageName, 0)
			PACKAGE_NAME = pi.packageName
			VERSION = pi.versionName.toString()
			VERSION_CODE = pi.versionCode.toString()
		}
		catch (e: Exception)
		{
			e.printStackTrace()
		}

		object : Thread()
		{
			override fun run()
			{
				val currentHandler = getDefaultUncaughtExceptionHandler()
				if (currentHandler !is DefaultExceptionHandler)
				{
					exceptionHandler = DefaultExceptionHandler()
					exceptionHandler!!.filesPath = filesPath
					exceptionHandler!!.handler = currentHandler
					setDefaultUncaughtExceptionHandler(exceptionHandler)
				}
			}
		}.start()
	}

	/**
	 * Search for stack trace files.
	 */
	public fun searchForStackTraces(): List<String>
	{
		val dir = File(filesPath)
		dir.mkdirs()
		val filter = FilenameFilter { _, name -> name.endsWith(".json") }
		return dir.list(filter)?.asList() ?: arrayListOf()
	}

	/**
	 * Forces a manual exception post
	 */
	public fun sendException(e: Exception)
	{
		exceptionHandler?.sendException(e, "CAUGHT EXCEPTION")
	}
}
